### ISSUE TITLE
pcapgo/read.go: Shouldn't ReadPacketData try to peek the packet header to check for corruption before reading?

### ISSUE THREAD

**Issue by assafmo** [google/gopacket#550]
In this code section:
https://github.com/google/gopacket/blob/a35e09f9f224786863ce609de910bc82fc4d4faf/pcapgo/read.go#L119-L140

I think maybe ReadPacketData should peek 16 bytes, check for corruption according to the [packet header specs](https://wiki.wireshark.org/Development/LibpcapFileFormat#Record_.28Packet.29_Header), and if there is corruption then skip 1 bytes.

My thinking is that maybe the next good packet header can start somewhere in the last 15 bytes of the 16 bytes read.

Right now, if there is corruption somewhere in the input file, pcapgo/read.go reads in increments of 16 bytes and might miss on good packets.

This of course will be much slower when dealing with a corrupted pcap file, but OTOH might find more good packets otherwise missed.

**Comment by assafmo** [comment]
[pcapfix also skips 16 bytes](https://github.com/Rup0rt/pcapfix/blob/20765dc4412f78633335169965000ecfcbb543ba/pcap.c#L506) just like `pcapgo/read.go`.

I'd love some explanation for this behavior.

**Comment by assafmo** [comment]
> The big problem here is that it is not possibly or really complicated to actually verify a pcap packet header (the file header can somewhat be verified a bit, since the first 4 bytes need to match one of magic numbers).

I guess you are right. We don't have enough data in the packet header to detect corruption. I wish there was magic number in each packet header. :disappointed: 

> First, there is no specification. As far as I know the pcap-format originated from the source code of libpcap and it is the reference implementation (the wireshark link you provided is just a documentation of pcap-files discovered in the wild and existing source code). This lead to several different variants (as also mentioned in the wireshark wiki link).

I did not know this. Thank you!



> This can lead to e.g. the timestamp field being nanosecond although microsecond is expected, or wrong snaplength (see
https://github.com/google/gopacket/blob/a35e09f9f224786863ce609de910bc82fc4d4faf/pcapgo/read.go#L162
)

LOL :rofl: I wrote this comment.

> > Right now, if there is corruption somewhere in the input file, pcapgo/read.go reads in increments of 16 bytes and might miss on good packets.
> 
> Actually it reads it in increments of 16 bytes + what the length field contains.

Actually it reads also what the length field contains only if it thinks the packet is ok:
https://github.com/google/gopacket/blob/a35e09f9f224786863ce609de910bc82fc4d4faf/pcapgo/read.go#L118-L130

> > [pcapfix also skips 16 bytes](https://github.com/Rup0rt/pcapfix/blob/20765dc4412f78633335169965000ecfcbb543ba/pcap.c#L506) just like `pcapgo/read.go`.
> > I'd love to get some explanation/logic behind this behavior.
> 
> `pcapgo/read.go` doesn't do any skipping. In `readPacketHeader` 16 bytes are read into `r.buf` (`r.buf` has a length of 16), because a pcap header is 16 bytes (8 bytes timestamp + 4 bytes CaptureLength + 4 bytes Length). This is followed by decoding said fields. Since the header is directly followed by the packet data, `ReadPacketData` copies `ci.CaptureLength` bytes to `data`. For the next packet the whole process needs to be done again.

You are right. By "skipping" I meant "saying the packet is corrupted and moves on to read the next packet".

**Comment by assafmo** [comment]
> I think the whole thing would slow down the most frequent case (having a non-corrupted pcap file) and overcomplicate the reading code.

Yes, this will complicate the code, but it will slow it down only in a corrupted pcap file.

**Comment by assafmo** [comment]
> [pcapfix also skips 16 bytes](https://github.com/Rup0rt/pcapfix/blob/20765dc4412f78633335169965000ecfcbb543ba/pcap.c#L506) just like `pcapgo/read.go`.
> 
> I'd love to get some explanation/logic behind this behavior.

Actually I was wrong. pcapfix checks byte by byte for the next valid packet header.

https://github.com/Rup0rt/pcapfix/issues/14#issuecomment-433843845

**Comment by notti** [comment]
Ah I had a look at the source code when I wrote my first answer, but couldn't that figure out from a fast glimpse.

> > I think the whole thing would slow down the most frequent case (having a non-corrupted pcap file) and overcomplicate the reading code.
> 
> Yes, this will complicate the code, but it will slow it down only in a corrupted pcap file.

Since you need the checks to find out if it is corrupted, it would slow also the non-corrupted case down. Checking length vs capture length (like you did in #551) definitely makes sense, and as far as I can remember, libpcap does the same.

Another possibility to do what you want (read broken pcaps), would be to provide some methods, with which a program using the API could achieve this goal. E.g. add a Seek method and then provide a program in the examples directory. Said program could then basically use ReadPacketData, and if it fails or if additional checks fail use Seek to go back + offset and then continue with the process.
